unixfs/io/pbdagreader.go

```diff
@@ -95,10 +95,27 @@ func (dr *PBDagReader) precalcNextBuf(ctx context.Context) error {
 	}
 
 	nxt, err := dr.promises[dr.linkPosition].Get(ctx)
-	if err != nil {
+	dr.promises[dr.linkPosition] = nil
+	switch err {
+	case nil:
+	case context.DeadlineExceeded, context.Canceled:
+		err = ctx.Err()
+		if err != nil {
+			return ctx.Err()
+		}
+		// In this case, the context used to *preload* the node has been canceled.
+		// We need to retry the load with our context and we might as
```

Comment:
Could we elaborate a bit more on this? I had the impression we're always using the same `Context` from the `PBDagReader` structure. In which case the local `ctx` won't be cancelled but the one from the `NodePromise` will be? The context from the promise comes from this function, maybe in another call (at it may have been preloaded before), but at which point can the global context from the DAG reader change?

Reply:
We're using the context from the call Read call. IMO, that's the correct context as it allows us to cancel reads and seek elsewhere (canceling any associated preloading as well).

Reply:
Sorry, scratch that. If the user is using `Read`, we use the reader's cotnext. If the user is manually calling `CtxReadFull` (a context aware read), we use the supplied context. I don't really think any additional comments will help as the context is just whatever gets passed in (and listing every possible context is just going to lead to comment rot when that changes).

Reply:
> (and listing every possible context is just going to lead to comment rot when that changes).

It seems to me there are only (or mainly) two possibilities. The most common (by far I think) is that the reader's context is used everywhere, so in fact there is only one context (and this part of the code isn't executed). The second, and less common,

> If the user is manually calling CtxReadFull

which is when this scenario takes place (if understand correctly).

Clarifying that distinction seems worth it, IMO.

Reply:
Fair enough.